drivers/serial/uart_bridge.c

```diff
@@ -0,0 +1,232 @@
+
+	const struct device *peer_dev =
+		cfg->peer_dev[uart_bridge_get_idx(dev, bridge_dev, false)];
+	struct uart_bridge_peer_data *own_data =
+		&data->peer[uart_bridge_get_idx(dev, bridge_dev, true)];
+
+	uint8_t *recv_buf;
+	int rb_len, recv_len;
+	int ret;
+
+	if (ring_buf_space_get(&own_data->rb) < RING_BUF_FULL_THRESHOLD) {
+		LOG_DBG("%s: buffer full: pause", dev->name);
+		uart_irq_rx_disable(dev);
+		own_data->paused = true;
+		return;
+	}
+
+	rb_len = ring_buf_put_claim(&own_data->rb, &recv_buf, RING_BUF_SIZE);
+	if (rb_len == 0) {
+		LOG_WRN("%s: ring_buf full", dev->name);
+		return;
+	}
+
+	recv_len = uart_fifo_read(dev, recv_buf, rb_len);
+	if (recv_len < 0) {
+		ring_buf_put_finish(&own_data->rb, 0);
+		LOG_ERR("%s: rx error: %d", dev->name, recv_len);
+		return;
+	}
+
+	ret = ring_buf_put_finish(&own_data->rb, recv_len);
+	if (ret < 0) {
+		LOG_ERR("%s: ring_buf_put_finish error: %d", dev->name, rb_len);
+		return;
+	}
+
+	uart_irq_tx_enable(peer_dev);
+}
+
+static void uart_bridge_handle_tx(const struct device *dev,
+				  const struct device *bridge_dev)
+{
+	const struct uart_bridge_config *cfg = bridge_dev->config;
+	struct uart_bridge_data *data = bridge_dev->data;
+
+	const struct device *peer_dev =
+		cfg->peer_dev[uart_bridge_get_idx(dev, bridge_dev, false)];
+	struct uart_bridge_peer_data *peer_data =
+		&data->peer[uart_bridge_get_idx(dev, bridge_dev, false)];
+
+	uint8_t *send_buf;
+	int rb_len, sent_len;
+	int ret;
+
+	rb_len = ring_buf_get_claim(&peer_data->rb, &send_buf, RING_BUF_SIZE);
+	if (rb_len == 0) {
+		LOG_DBG("%s: buffer empty, disable tx irq", dev->name);
+		uart_irq_tx_disable(dev);
+		return;
+	}
+
+	sent_len = uart_fifo_fill(dev, send_buf, rb_len);
+	if (sent_len < 0) {
+		ring_buf_get_finish(&peer_data->rb, 0);
+		LOG_ERR("%s: tx error: %d", dev->name, sent_len);
+		return;
+	}
+
+	ret = ring_buf_get_finish(&peer_data->rb, sent_len);
+	if (ret < 0) {
+		LOG_ERR("ring_buf_get_finish error: %d", ret);
+		return;
+	}
+
+	if (peer_data->paused &&
+	    ring_buf_space_get(&peer_data->rb) > RING_BUF_FULL_THRESHOLD) {
+		LOG_DBG("%s: buffer free: resume", dev->name);
+		uart_irq_rx_enable(peer_dev);
+		peer_data->paused = false;
+		return;
+	}
+}
+
+static void interrupt_handler(const struct device *dev, void *user_data)
+{
+	const struct device *bridge_dev = user_data;
+
+	while (uart_irq_update(dev) && uart_irq_is_pending(dev)) {
+		if (uart_irq_rx_ready(dev)) {
+			uart_bridge_handle_rx(dev, bridge_dev);
+		}
+		if (uart_irq_tx_ready(dev)) {
+			uart_bridge_handle_tx(dev, bridge_dev);
+		}
+	}
+}
+
+static int uart_bridge_init(const struct device *dev)
+{
+	const struct uart_bridge_config *cfg = dev->config;
+	struct uart_bridge_data *data = dev->data;
+
+	ring_buf_init(&data->peer[0].rb, RING_BUF_SIZE, data->peer[0].buf);
+	ring_buf_init(&data->peer[1].rb, RING_BUF_SIZE, data->peer[1].buf);
+
+	/* Wait 100ms for the host to do all settings */
+	k_msleep(100);
```

Comment:
What does this do exactly? Maybe you need some kind of enable/disable instead? If needed, it could be turned into a property of this (virtual) device?